### PR TITLE
fix: add masterSubnet to azuredeploy.parameters.json while using custom VNET

### LINKED ISSUE #210: Windows nodes initialization script fails in hybrid cluster with custom VNet
**Is this a request for help?**:
No
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
ISSUE
---

**What version of aks-engine?**:
Version: v0.28.1
GitCommit: efc9dcfd
GitTreeState: clean
---

**Kubernetes version**:
1.11.5

**What happened**:
Deployed hybrid cluster in custom vnet with one subnet. kuberneteswindowssetup.ps1 failed to configure networking because MasterSubnet parameter was null or empty. 

Fixed this manually by providing 'masterSubnet' parameter in azuredeploy.parameters.json and deploying generated ARM template.

Cluster definition:
```json
{
    "apiVersion": "vlabs",
    "properties": {
        "orchestratorProfile": {
            "orchestratorType": "Kubernetes",
            "orchestratorRelease": "1.11",
            "orchestratorVersion": "1.11.5"
        },
        "masterProfile": {
            "count": 1,
            "dnsPrefix": "dnsPrefix",
            "vmSize": "Standard_D2_v2",
            "vnetSubnetId": "subnet",
            "agentVnetSubnetId": "subnet",
            "firstConsecutiveStaticIP": "192.168.255.200",
            "vnetCidr": "192.168.255.0/24"
        },
        "agentPoolProfiles": [{
                "name": "win",
                "count": 1,
                "vmSize": "Standard_DS2_v2",
                "vnetSubnetId": "subnet",
                "osType": "Windows"
            },
            {
                "name": "linux",
                "count": 1,
                "vmSize": "Standard_DS2_v2",
                "vnetSubnetId": "subnet"
            }
        ],
        "windowsProfile": {
            "adminUsername": "adminUsername",
            "adminPassword": "adminPassword"
        },
        "linuxProfile": {
            "adminUsername": "adminUsername",
            "ssh": {
                "publicKeys": [{
                    "keyData": "ssh-rsa ..."
                }]
            }
        },
        "servicePrincipalProfile": {
            "clientId": "clientId",
            "secret": "secret"
        }
    }
}
```

**What you expected to happen**:
masterSubnet parameter is set during ARM template generation process

**How to reproduce it** (as minimally and precisely as possible):
Deploy provided cluster into a custom VNet with a single subnet.

**Anything else we need to know**:

### LINKED ISSUE #341: Windows pod creation always fails with Pod sandbox changed - Failed create pod sandbox
**Is this a request for help?**:

---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):

Issue

**What version of aks-engine?**:

Version: v0.28.1
GitCommit: efc9dcfd
GitTreeState: clean

**Kubernetes version**: v1.9.11

**What happened**: 

Our containerized application suddenly stopped working and now all pods on windows nodes are stuck in a restart cycle of:

Normal   SandboxChanged          5m (x2207 over 4h)   kubelet, 3401k8s012  Pod sandbox changed, it will be killed and re-created.

Warning  FailedCreatePodSandBox  20s (x2230 over 4h)  kubelet, 3401k8s012  Failed create pod sandbox.

**What you expected to happen**:

I have deleted all deployments and services, rebooted all Windows nodes and the Linux master and attempted to reapply all deployments and services, but all pods are stuck in the error cycle above

**How to reproduce it** (as minimally and precisely as possible):

I don't know the exact steps as it occurred suddenly. However this has now happened in two separate aks-engine deployments in the last 4 days.

**Anything else we need to know**:

It's a mixed cluster of Windows and Linux nodes. This only effects pods running on Windows nodes; however it impacts all pods on Windows nodes.

The cluster looks like:

PS:> kubectl get nodes -o wide
NAME                        STATUS    ROLES     AGE       VERSION   EXTERNAL-IP   OS-IMAGE                    KERNEL-VERSION     CONTAINER-RUNTIME
3401k8s010                  Ready     agent     7d        v1.9.11   <none>        Windows Server Datacenter   10.0.17763.253     docker://18.9.0
3401k8s011                  Ready     agent     7d        v1.9.11   <none>        Windows Server Datacenter   10.0.17763.253     docker://18.9.0
3401k8s012                  Ready     agent     7d        v1.9.11   <none>        Windows Server Datacenter   10.0.17763.253     docker://18.9.0
3401k8s013                  Ready     agent     7d        v1.9.11   <none>        Windows Server Datacenter   10.0.17763.253     docker://18.9.0
3401k8s014                  Ready     agent     7d        v1.9.11   <none>        Windows Server Datacenter   10.0.17763.253     docker://18.9.0
k8s-linuxpool1-34014015-0   Ready     agent     7d        v1.9.11   <none>        Ubuntu 16.04.5 LTS          4.15.0-1035-azure  docker://3.0.1
k8s-master-34014015-0       Ready     master    7d        v1.9.11   <none>        Ubuntu 16.04.5 LTS          4.15.0-1036-azure  docker://3.0.1

Deployed with aks-version v0.28.1

PS:> aks-engine version
Version: v0.28.1
GitCommit: efc9dcfd
GitTreeState: clean

### PR DESCRIPTION
**Pull request by zhiweiv**
**Reason for Change**:

masterSubnet is used by Windows nodes, it is missing from azuredeploy.parameters.json while using custom VNET, we can add it to azuredeploy.parameters.json manually while building cluster as a workaround, however while scaling cluster, aks-engine updates the cluster directly, we don't get change to update the template, the new added Windows nodes do not join the cluster due to missing masterSubnet. This PR make it work at least while master and agents using same custom VNET.

**Issue Fixed**:

Fixes #210 
Fixes #341 

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: pkg/engine/params.go [hunk 1]
```diff
@@ -62,9 +62,8 @@ func getParameters(cs *api.ContainerService, generatorCode string, aksEngineVers
 			if properties.MasterProfile.IsVirtualMachineScaleSets() {
 				addValue(parametersMap, "agentVnetSubnetID", properties.MasterProfile.AgentVnetSubnetID)
 			}
-			if properties.OrchestratorProfile.IsKubernetes() {
-				addValue(parametersMap, "vnetCidr", properties.MasterProfile.VnetCidr)
-			}
+			addValue(parametersMap, "vnetCidr", properties.MasterProfile.VnetCidr)
```
[the remark this answers is not in the document]
Comment (zhiweiv) [reply pkg/engine/params.go]: k8s is the only supported orchestrator now, so I guess the if condition is not necessary any more.  Do I need to revert it back?

[the remark this answers is not in the document]
Comment (zhiweiv) [reply pkg/engine/params.go]: Updated

File: pkg/engine/params.go [hunk 2]
```diff
@@ -64,6 +64,7 @@ func getParameters(cs *api.ContainerService, generatorCode string, aksEngineVers
 			}
 			if properties.OrchestratorProfile.IsKubernetes() {
 				addValue(parametersMap, "vnetCidr", properties.MasterProfile.VnetCidr)
+				addValue(parametersMap, "masterSubnet", properties.MasterProfile.VnetCidr)
```
Comment (tariq1890) [inline pkg/engine/params.go]: Why are we not adding this in `params_k8s.go` ?

Comment (tariq1890) [reply pkg/engine/params.go]: Note to self: Check the pre-conditions for populating `VnetCidr`

Comment (zhiweiv) [reply pkg/engine/params.go]:
```go
if properties.MasterProfile.IsCustomVNET() {
	addValue(parametersMap, "masterVnetSubnetID", properties.MasterProfile.VnetSubnetID)
	if properties.MasterProfile.IsVirtualMachineScaleSets() {
		addValue(parametersMap, "agentVnetSubnetID", properties.MasterProfile.AgentVnetSubnetID)
	}
	if properties.OrchestratorProfile.IsKubernetes() {
		addValue(parametersMap, "vnetCidr", properties.MasterProfile.VnetCidr)
		addValue(parametersMap, "masterSubnet", properties.MasterProfile.VnetCidr)
	}
} else {
	addValue(parametersMap, "masterSubnet", properties.MasterProfile.Subnet)
	addValue(parametersMap, "agentSubnet", properties.MasterProfile.AgentSubnet)
}
```
This is the whole part of related code, masterSubnet is added in else condition, I followed this rule to add my code here. Do I need to move it to params_k8s.go?

Comment (zhiweiv) [reply pkg/engine/params.go]: The **vnetCidr** is not used in any condition now.

**Comment by PatrickLang** [comment]
/LGTM

@jsturtevant - do you have anyone looking for this functionality and if so would you be able to help document this?

File: pkg/engine/params.go (hunk 2 above)
Comment (tariq1890) [reply pkg/engine/params.go]: Then I believe we should change that to use VnetCidr, not duplicate values across different parameters.

Comment (zhiweiv) [reply pkg/engine/params.go]: There are a few parameters related to custom VNET, they are  `vnetCidr`/`kubeClusterCidr`/`masterSubnet`/`agentSubnet`. `vnetCidr` is not used, `agentSubnet` is a null value. `kubeClusterCidr` and `masterSubnet` are added to azure cni's exception list only.

I guess there were some thoughts for custom VNET, but not fully implemented yet on Windows. From my side this PR is more as a workaround with minimal change to make things work. 

@PatrickLang  
Any ideas?

**Comment by zhiweiv** [comment]
@tariq1890 @PatrickLang @CecileRobertMichon 
Do we have change to merge this to trunk, currently I am using local build with changes from this PR. It is inconvenient because I have to rebuild often to catch up new version.

**Comment by CecileRobertMichon** [comment]
@tariq1890 what is still holding this PR?

**Comment by tariq1890** [comment]
I am enquiring if there is a way instead of going for a workaround here. Having values duplicated across multiple parameters seems like a code/design smell to me.

**Comment by zhiweiv** [comment]
The PR maybe not a good solution, I will keep using local build as workaround. AKS-Engine doesn't work for Windows nodes with custom VNET by default now, hope this can be fixed sometime.

**Comment by rogersollberger** [comment]
Is it true, this fix is still not in the current version (v0.35.1) ? So there is currently no way to use a hybrid setup with custom vnet than to manually add the mastersubnet after aks-engine generate to azuredeploy.parameters.json?

**Comment by zhiweiv** [comment]
@rogersollberger 
You can fork and modify the code like this PR on your branch, then use your own build of AKS-Engine.

**Pull request closed without merging**